Add latency detector for desktop sessions

probakowski opened the pull request:
This PR adds latency measurement for Windows Desktop session by calculating latency from browser to proxy and from proxy to the desktop.

![latency_detector](https://github.com/user-attachments/assets/f081d80b-e311-49a0-97d6-996c7c343cc9)
op.

changelog: Add latency detector for desktop sessions


Branched off of https://github.com/gravitational/teleport/pull/40417

zmb3 committed:
wip

probakowski committed:
Merge remote-tracking branch 'origin/master' into zmb3/desktop-latency-detector

probakowski committed:
add ping message and latency from desktop side

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1631,6 +1636,39 @@ func decodeSharedDirectoryTruncateResponse(in io.Reader) (SharedDirectoryTruncat
 	return res, err
 }
 
+type LatencyStats struct {
```
Let's add godocs.

```suggestion
// LatencyStats is used to report the latency of the
// connection(s) to the web UI.
type LatencyStats struct {
```

zmb3 commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -435,6 +436,15 @@ func (c *Client) startInputStreaming(stopCh chan struct{}) error {
 			return err
 		}
 
+		if m, ok := msg.(tdp.Ping); ok {
```
Do we need a ping message at all? What if the agent just did periodic latency checks and reported its status by injecting a `LatencyStats` message in the stream?

probakowski replied:
We need it for measuring the latency between proxy and WDS, otherwise we will measure misleading latency where WDS is close to the desktop but far away from the proxy

zmb3 reviewed (commented):
We also need to consider compatibility in both directions:

1. If proxy is ahead of agents, and proxy knows about the new messages but agents don't, then the proxy should either hide latency detector completely, or show only the first leg of the connection. (We can ask design for ideas how to indicate this in the UI).
2. If agents are ahead of proxy, then the proxy should just drop latency messages and not crash on an unknown message type.

Please try both use cases and let me know what your plan for addressing them is.

probakowski committed:
version

probakowski committed:
Fix backward compatibility

probakowski committed:
godocs

probakowski committed:
formatting

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/desktop-latency-detector

probakowski committed:
fix imports

probakowski committed:
formatting

probakowski committed:
formatting

probakowski committed:
log and gci

probakowski committed:
lint

gzdunek committed:
Apply tooltip on the icon directly, instead of on the Menu component

This fixes a problem where onMouseLeave in HoverTooltip wasn't called and the tooltip didn't disappear.

gzdunek committed:
Use consistent spacing between top bar elements

gzdunek commented:
I pushed a fix for the tooltip not disappearing when the popup got closed. The problem occurred because the tooltip was applied to the entire popup instead of just the icon. I suspect the modal, which was the child of the tooltip, prevented its `onMouseLeave` handler from being triggered. 
Applying a tooltip to a menu feels incorrect anyways, so I moved it to the icon instead.

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/desktop-latency-detector

probakowski committed:
updates from origin

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/desktop-latency-detector

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/desktop-latency-detector

probakowski committed:
e

probakowski committed:
e

probakowski committed:
lint

probakowski commented:
@zmb3 I've tested backward compatibility in both directions
- if proxy is ahead of WDS I've added version check and we won't send `Ping` messages avoiding error, we won't show the latency detector UI at all in such case
- if WDS is ahead of proxy it will never send `Ping` message as well because it does it only in response to the message received from proxy

gzdunek commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1631,6 +1636,44 @@ func decodeSharedDirectoryTruncateResponse(in io.Reader) (SharedDirectoryTruncat
 	return res, err
 }
 
+// LatencyStats is used to report the latency of the connection(s) to the web UI.
```
Soon we will have desktop access in Connect (I assume we will add the latency monitor there at some point), so I’d avoid referring specifically to the Web UI and instead use something like 'client' instead.

probakowski replied:
Yep, changed

gzdunek commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1631,6 +1636,44 @@ func decodeSharedDirectoryTruncateResponse(in io.Reader) (SharedDirectoryTruncat
 
+// LatencyStats is used to report the latency of the connection(s) to the web UI.
+type LatencyStats struct {
+	BrowserLatency uint32
+	DesktopLatency uint32
```
What do you think of calling these fields `Client` and `Server`? 
Both `latency.Statistics` and the frontend's `LatencyDiagnostic` already use these terms, so keeping them consistent might be more convenient (we wouldn't have to convert the object on the JS side).

probakowski replied:
Makes sense, renamed

gzdunek commented on web/packages/teleport/src/DesktopSession/DesktopSession.tsx:
```diff
@@ -156,6 +161,24 @@ export function DesktopSession({
     addAlert,
   } = useDesktopSession(client, aclAttempt);
 
+  const [latencyStats, setLatencyStats] = useState(undefined);
```
Missing type.
```suggestion
  const [latencyStats, setLatencyStats] = useState<Latency | undefined>();
```

gzdunek commented on web/packages/teleport/src/DesktopSession/DesktopSession.tsx:
```diff
@@ -156,6 +161,24 @@ export function DesktopSession({
   } = useDesktopSession(client, aclAttempt);
 
+  const [latencyStats, setLatencyStats] = useState(undefined);
+  useEffect(() => {
```
The client was recently refactored, allowing us to subscribe to events in a more concise way.

```diff
diff --git a/web/packages/teleport/src/DesktopSession/DesktopSession.tsx b/web/packages/teleport/src/DesktopSession/DesktopSession.tsx
index f1d4d9dd99..d1c10e7c5e 100644
--- a/web/packages/teleport/src/DesktopSession/DesktopSession.tsx
+++ b/web/packages/teleport/src/DesktopSession/DesktopSession.tsx
@@ -27,6 +27,7 @@ import Dialog, {
   DialogHeader,
   DialogTitle,
 } from 'design/Dialog';
+import { Latency } from 'shared/components/LatencyDiagnostic';
 import { Attempt, makeSuccessAttempt, useAsync } from 'shared/hooks/useAsync';
 
 import { useTeleport } from 'teleport';
@@ -35,12 +36,7 @@ import TdpClientCanvas from 'teleport/components/TdpClientCanvas';
 import cfg, { UrlDesktopParams } from 'teleport/config';
 import { KeyboardHandler } from 'teleport/DesktopSession/KeyboardHandler';
 import { AuthenticatedWebSocket } from 'teleport/lib/AuthenticatedWebSocket';
-import {
-  ButtonState,
-  ScrollAxis,
-  TdpClient,
-  TdpClientEvent,
-} from 'teleport/lib/tdp';
+import { ButtonState, ScrollAxis, TdpClient } from 'teleport/lib/tdp';
 import { useListener } from 'teleport/lib/tdp/client';
 import { shouldShowMfaPrompt, useMfaEmitter } from 'teleport/lib/useMfa';
 import { getHostName } from 'teleport/services/api';
@@ -161,24 +157,6 @@ export function DesktopSession({
     addAlert,
   } = useDesktopSession(client, aclAttempt);
 
-  const [latencyStats, setLatencyStats] = useState(undefined);
-  useEffect(() => {
-    if (!client) {
-      return;
-    }
-    const setStats = stats => {
-      console.log('got latency', stats);
-      setLatencyStats({
-        client: stats.browserLatency,
-        server: stats.desktopLatency,
-      });
-    };
-    client.on(TdpClientEvent.LATENCY_STATS, setStats);
-    return () => {
-      client.removeListener(TdpClientEvent.LATENCY_STATS, setStats);
-    };
-  }, [client]);
-
   const [tdpConnectionStatus, setTdpConnectionStatus] =
     useState<TdpConnectionStatus>({ status: '' });
 
@@ -214,6 +192,17 @@ export function DesktopSession({
     setTimeout(() => tdpClientCanvasRef.current?.focus());
   }, []);
 
+  const [latencyStats, setLatencyStats] = useState<Latency | undefined>();
+  useListener(
+    client.onLatencyStats,
+    useCallback(stats => {
+      setLatencyStats({
+        client: stats.browserLatency,
+        server: stats.desktopLatency,
+      });
+    }, [])
+  );
+
   useListener(client.onClipboardData, onClipboardData);
 
   const handleFatalError = useCallback(
diff --git a/web/packages/teleport/src/DesktopSession/TopBar.tsx b/web/packages/teleport/src/DesktopSession/TopBar.tsx
index 4d00e01dbd..fb56e3f2fd 100644
--- a/web/packages/teleport/src/DesktopSession/TopBar.tsx
+++ b/web/packages/teleport/src/DesktopSession/TopBar.tsx
@@ -21,7 +21,10 @@ import { useTheme } from 'styled-components';
 import { Flex, Text, TopNav } from 'design';
 import { Clipboard, FolderShared } from 'design/Icon';
 import { HoverTooltip } from 'design/Tooltip';
-import { LatencyDiagnostic } from 'shared/components/LatencyDiagnostic';
+import {
+  Latency,
+  LatencyDiagnostic,
+} from 'shared/components/LatencyDiagnostic';
 import type { NotificationItem } from 'shared/components/Notification';
 
 import ActionMenu from './ActionMenu';
@@ -113,8 +116,5 @@ type Props = {
   onCtrlAltDel: VoidFunction;
   alerts: NotificationItem[];
   onRemoveAlert(id: string): void;
-  latency: {
-    client: number;
-    server: number;
-  };
+  latency: Latency;
 };
diff --git a/web/packages/teleport/src/lib/tdp/client.ts b/web/packages/teleport/src/lib/tdp/client.ts
index 49cca4a535..19383779ea 100644
--- a/web/packages/teleport/src/lib/tdp/client.ts
+++ b/web/packages/teleport/src/lib/tdp/client.ts
@@ -31,6 +31,7 @@ import { MfaChallengeResponse } from 'teleport/services/mfa';
 
 import Codec, {
   FileType,
+  LatencyStats,
   MessageType,
   PointerData,
   Severity,
@@ -224,6 +225,11 @@ export default class Client extends EventEmitterMfaSender {
     return () => this.off(TdpClientEvent.TDP_CLIENT_SCREEN_SPEC, listener);
   };
 
+  onLatencyStats = (listener: (stats: LatencyStats) => void) => {
+    this.on(TdpClientEvent.LATENCY_STATS, listener);
+    return () => this.off(TdpClientEvent.LATENCY_STATS, listener);
+  };
+
   private async initWasm() {
     // select the wasm log level
     let wasmLogLevel = LogType.OFF;
```

probakowski replied:
Thanks, I applied you suggestions (some code moved after other changes), could you check if I did it correctly?

gzdunek replied:
We only lost a `gap` property, I left a comment. Everything else is correct 👍

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/desktop-latency-detector

probakowski committed:
rework UI after merge

probakowski committed:
Rename fields in backend

probakowski committed:
Merge branch 'master' into probakowski/desktop-latency-detector

probakowski committed:
prettier

gzdunek reviewed (approved):
With a few comments.

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/desktop-latency-detector

probakowski committed:
Update web/packages/shared/components/DesktopSession/TopBar.tsx

Co-authored-by: Grzegorz Zdunek <gzdunek@users.noreply.github.com>

probakowski committed:
review comments

probakowski commented on web/packages/shared/libs/tdp/codec.ts:
```diff
@@ -323,6 +324,12 @@ export enum FileType {
+// | message type (35) | browser_latency uint32 | desktop_latency uint32 |
+export type LatencyStats = {
+  browserLatency: number;
+  desktopLatency: number;
```
Done!

probakowski committed:
Merge branch 'master' into probakowski/desktop-latency-detector

probakowski committed:
fix ui

probakowski committed:
add env var to disable windows desktop "ping"

probakowski committed:
review comment

probakowski committed:
review comment

zmb3 commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -433,6 +438,17 @@ func (c *Client) startInputStreaming(stopCh chan struct{}) error {
+					conn.Close()
+				}
+			}
+			c.cfg.Conn.WriteMessage(m)
```
What is the `WriteMessage` for? We receive a  `tdp.Ping` message and then we send the exact same message back?

probakowski replied:
Yes, to avoid adding yet another tdp message like `Pong`

zmb3 replied:
Ok, I would add a brief comment:

```
// Upon receiving a ping message, we make a connection
// to the host and send the same message back to the proxy.
// The proxy will then compute the round trip time.
```

Also, should we have an error check on this `WriteMessage` call?

zmb3 replied:
@probakowski I'm still waiting a response on this one.

probakowski replied:
I've added comment and added check for error

zmb3 commented on lib/web/desktop.go:
```diff
@@ -539,18 +542,126 @@ func (c *connector) tryConnect(ctx context.Context, clusterName, desktopServiceI
+	}
+	for {
+		select {
+		case p := <-d.ch:
```
```suggestion
		case pong := <-d.ch:
```

This will make it clear that it's the response to the ping.

probakowski committed:
Update lib/web/desktop.go

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

probakowski committed:
Refactor latency monitoring

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/desktop-latency-detector

probakowski committed:
fix spelling

probakowski committed:
Merge branch 'master' into probakowski/desktop-latency-detector

probakowski committed:
remove monitorSessionLatency

probakowski committed:
gci

zmb3 commented on lib/web/desktop.go:
```diff
@@ -539,18 +542,108 @@ func (c *connector) tryConnect(ctx context.Context, clusterName, desktopServiceI
 // Desktop Serivce (wds)
-func proxyWebsocketConn(ws *websocket.Conn, wds net.Conn) error {
+func proxyWebsocketConn(ws *websocket.Conn, wds net.Conn, log *slog.Logger, version string) error {
+	ctx, cancel := context.WithCancel(context.Background())
```
Should we pass in the context from `createDekstopConnection` rather than using a background context here?

probakowski replied:
We should, I've added it

zmb3 commented on lib/web/desktop.go:
```diff
@@ -539,18 +542,108 @@ func (c *connector) tryConnect(ctx context.Context, clusterName, desktopServiceI
+	tdpMessagesToSend := make(chan tdp.Message)
+	errs := make(chan error, 3)
+
+	latencySupported, err := utils.MinVerWithoutPreRelease(version, "18.0.0")
```
Let's make this 17.5.0?

probakowski replied:
I thought we merge it here with `18.0.0`, backport to v17 with `17.5.0` and then do another PR with change to `17.5.0` in master. WDYT?

probakowski replied:
I've changed it to v17.5, we can change it before v18 release if needed

zmb3 commented on lib/web/desktop.go:
```diff
@@ -539,18 +542,108 @@ func (c *connector) tryConnect(ctx context.Context, clusterName, desktopServiceI
 
-	errs := make(chan error, 2)
+	tdpMessagesToSend := make(chan tdp.Message)
+	errs := make(chan error, 3)
```
I wonder if it would be better to use an `errgroup` at this point?

probakowski replied:
I think it's good idea - it should be safer and leads to more idiomatic code. I've changed it

probakowski committed:
review comment

probakowski committed:
review comment

zmb3 reviewed (approved):
LGTM other than the v18/v17.5 decision.

Also, please update the RFD with all the TDP messages to account for the new latency stats and ping messages.

probakowski committed:
Update RFD and version

probakowski committed:
Merge branch 'master' into probakowski/desktop-latency-detector

probakowski committed:
fix gaps

probakowski committed:
Merge branch 'master' into probakowski/desktop-latency-detector

espadolini commented on lib/web/desktop.go:
```diff
@@ -535,19 +539,108 @@ func (c *connector) tryConnect(ctx context.Context, clusterName, desktopServiceI
-	errs := make(chan error, 2)
+	tdpMessagesToSend := make(chan tdp.Message)
+
+	latencySupported, err := utils.MinVerWithoutPreRelease(version, "17.5.0")
```
Instead of relying on indirect info such as the version string (in a heartbeat, even, not even from the actual connection in use) can we start thinking about exchanging capabilities as part of the TDP "handshake", so to speak?

probakowski commented on lib/web/desktop.go:
```diff
@@ -535,19 +539,108 @@ func (c *connector) tryConnect(ctx context.Context, clusterName, desktopServiceI
+	// run a goroutine to pick TDP messages up from a channel and send
+	// them to the browser
+	errs.Go(func() error {
+		for msg := range tdpMessagesToSend {
```
It does, could you please check https://github.com/gravitational/teleport/pull/54604/commits/0551d440ad5261ee687552b4bf4e02e0ae53d011 where I tried to fix all 3 problems you mention
